spring-cloud-gcp-pubsub-stream-binder/src/test/java/org/springframework/cloud/gcp/stream/binder/pubsub/PubSubEmulator.java

```diff
@@ -0,0 +1,216 @@
+			waitForConfigCreation();
+		}
+
+		Process envInitProcess = new ProcessBuilder("gcloud", "beta", "emulators", "pubsub", "env-init").start();
+
+		// env-init output is a shell command of the form "export
+		// PUBSUB_EMULATOR_HOST=localhost:8085".
+		String emulatorInitString = new BufferedReader(new InputStreamReader(envInitProcess.getInputStream()))
+				.readLine();
+		envInitProcess.waitFor();
+		String emulatorHostPort = emulatorInitString.substring(emulatorInitString.indexOf('=') + 1);
+		extractTeardownParams(emulatorHostPort);
+
+		this.binder = new PubSubTestBinder(emulatorHostPort);
+	}
+
+	/**
+	 * Shuts down the two emulator processes.
+	 *
+	 * gcloud command is shut down through the direct process handle. java process is
+	 * identified and shut down through shell commands.
+	 *
+	 * There should normally be only one process with that host/port combination, but if there
+	 * are more, they will be cleaned up as well.
+	 */
+	@Override
+	protected void after() {
+		try {
+			String hostPortParams = String.format("--host=%s --port=%s", this.emulatorHost, this.emulatorPort);
+			Process psProcess = new ProcessBuilder("ps", "-v").start();
+			new BufferedReader(new InputStreamReader(psProcess.getInputStream()))
+					.lines()
+					.filter(psLine -> psLine.contains(hostPortParams))
+					.map(psLine -> new StringTokenizer(psLine).nextToken())
+					.forEach(pid -> {
+						this.killProcess(pid);
+					});
+		}
+		catch (IOException e) {
+			LOGGER.error("Failed to cleanup: ", e);
+		}
+		this.emulatorProcess.destroy();
+	}
+
+	/**
+	 * Returns an already-configured emulator binder when called from within a JUnit method.
+	 */
+	public PubSubTestBinder getBinder() {
+		return this.binder;
+	}
+
+	/**
+	 * Waits until a PubSub emulator configuration file is present.
+	 *
+	 * Fails if the file does not appear after 10 seconds.
+	 *
+	 * @throws InterruptedException which should interrupt the peaceful slumber and bubble up
+	 * to fail the test.
+	 */
+	private void waitForConfigCreation() throws InterruptedException {
+		int attempts = 10;
+		while (!Files.exists(this.EMULATOR_CONFIG_PATH) && --attempts >= 0) {
+			Thread.sleep(1000);
+		}
+		if (attempts < 0) {
+			throw new RuntimeException(
+					"Emulator could not be configured due to missing env.yaml. Are PubSub and beta tools installed?");
+		}
+	}
+
+	/**
+	 * Waits until a PubSub emulator configuration file is updated.
+	 *
+	 * Fails if the file does not update after 1 second.
+	 *
+	 * @throws InterruptedException which should interrupt the peaceful slumber and bubble up
+	 * to fail the test.
+	 */
+	private void waitForConfigUpdate(WatchService watchService) throws InterruptedException {
+		int attempts = 10;
+		while (--attempts >= 0) {
+			WatchKey key = watchService.poll(100, TimeUnit.MILLISECONDS);
+
+			if (key != null) {
+				Optional<Path> configFilePath = key.pollEvents().stream().filter(
+						event -> event.kind() == StandardWatchEventKinds.ENTRY_MODIFY)
```

Comment:
I believe we don't need filter here since we are subscribing only for the `StandardWatchEventKinds.ENTRY_MODIFY`